File: behavioral/typescript/command_structured/invoker.ts
import { Command } from './interfaces';

export class RemoteControl {
  private commandHistory: Command[] = [];
  private undoStack: Command[] = [];
  
  setCommand(command: Command): void {
    this.commandHistory.push(command);
  }
  
  buttonPressed(slot: number): void {
    if (slot < this.commandHistory.length) {
      this.commandHistory[slot].execute();
      this.undoStack.push(this.commandHistory[slot]);
    }
  }
  
  undoButtonPressed(): void {
    if (this.undoStack.length > 0) {
      const command = this.undoStack.pop()!;
      command.undo();
    } else {
      console.log('Nothing to undo');
    }
  }
}

File: behavioral/dart/command_structured/invoker.dart
import 'interfaces.dart';

class RemoteControl {
  final List<Command> _commandHistory = [];
  final List<Command> _undoStack = [];
  
  void setCommand(Command command) {
    _commandHistory.add(command);
  }
  
  void buttonPressed(int slot) {
    if (slot < _commandHistory.length) {
      _commandHistory[slot].execute();
      _undoStack.add(_commandHistory[slot]);
    }
  }
  
  void undoButtonPressed() {
    if (_undoStack.isNotEmpty) {
      final command = _undoStack.removeLast();
      command.undo();
    } else {
      print('Nothing to undo');
    }
  }
}
